### ISSUE TITLE
SVD is slow on GPU vs CPU for skinny matrices

### ISSUE THREAD

**Issue by n-gao** [pytorch/pytorch#41306]
## 🐛 Bug
Performing SVD on the GPU is extremely slow and as far as I it is an open research quetion whether SVD in general can gain a lot by performing it on the GPU. So I want to propose to execute it on the CPU by default.

## To Reproduce
CPU:
```python
A = torch.randn(100, 10, 10).cpu()
%timeit torch.svd(A); torch.cuda.synchronize()
```
Result:
```
2.02 ms ± 71.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
CUDA:
```python
A = torch.randn(100, 10, 10).cuda()
%timeit torch.svd(A); torch.cuda.synchronize()
```
Result:
```
143 ms ± 4.02 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
Here is a figure illustrating the behavior depending on the batch_size:
![image](https://user-images.githubusercontent.com/9084670/87221301-3701c000-c36b-11ea-8913-97d48eaa25ca.png)

## Expected behavior
The CUDA result should be at least as good as the CPU otherwise there is no point in using CUDA.

## Environment
 - PyTorch Version (e.g., 1.0): 1.5
 - OS (e.g., Linux): Windows 10 2004
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.7.4
 - CUDA/cuDNN version: 11.0
 - GPU models and configuration: Nvidia GTX 1050
 - Any other relevant information:

## Additional context
I am aware that I could manually move the tensor to the CPU before computing the SVD, but this has several drawbacks:
1. It is unintuitive that the CUDA version of SVD is worse than the CPU version
2. Functions like `torch.slogdet` use SVD internally during the backward pass if the tensor is singular. This can only be avoided by implementing custom gradients or by performing `slogdet` on the CPU despite `slogdet` being faster with CUDA.

cc @vincentqb @vishwakftw @SsnL @jianyuh @VitalyFedyunin @ngimel

**Comment by vishwakftw** [comment]
Relevant issue: https://github.com/pytorch/pytorch/issues/23940 - filed over a year ago.

**Comment by ngimel** [comment]
cc @ptrblck

**Comment by ngimel** [comment]
Should be somewhat alleviated (on cuda11) when we merge #48436, but agreed, magma implementation is hopeless (and calls a million of cudaMalloc and cudaFree to boot). cc @heitorschueroff

**Comment by ngimel** [comment]
cc @xwang233, @heitorschueroff

**Comment by xwang233** [comment]
CUDA SVD only has performance better than CPU at large matrix size, or large batch size + small matrix size. To use a faster cusolver batched implementation, the matrix size needs to be <= 32. With matrix size 40, or say 33, the current GPU implementation has to launch many small CUDA kernels in serial for loop, which is very inefficient. There is barely anything we can do at pytorch side. If possible, I would suggest using batched matrix with matrix size <= 32.

I also tried 1.7 with MAGMA backend, the GPU performance is within 10% difference from the current 1.8 cusolver backend.

**Comment by tingxueronghua** [comment]
> CUDA SVD only has performance better than CPU at large matrix size, or large batch size + small matrix size. To use a faster cusolver batched implementation, the matrix size needs to be <= 32. With matrix size 40, or say 33, the current GPU implementation has to launch many small CUDA kernels in serial for loop, which is very inefficient. There is barely anything we can do at pytorch side. If possible, I would suggest using batched matrix with matrix size <= 32.
> 
> I also tried 1.7 with MAGMA backend, the GPU performance is within 10% difference from the current 1.8 cusolver backend.

Thanks for your answers! But I still have a question. I hope you could help me with it!
If I implement the code by myself with CUDA C, trying to get SVD decompostion of matrices in shape(60, 3) and batchsize more than 1000, is there any way more efficient to calculate? Under this condition, I found the SVD in cuSolver extremely slower than numpy version. Is there anything related with the serial calculation feature described in your answer? By the way, I use  cusolverDnSgesvdj in cuSolver.
I noticed you mentioned "large batchsize + small matrix size", so I am wondering is there anything I could improve in this situation.

**Comment by xwang233** [comment]
@tingxueronghua , for your use case (60x3 matrix), cusolver provides a function named gesvdaStridedBatched (https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverDN-lt-t-gt-gesvda), which is optimized for tall skinny matrix. You may want to try that in your CUDA program, and see if it gives better performance.

**Comment by tingxueronghua** [comment]
> @tingxueronghua , for your use case (60x3 matrix), cusolver provides a function named gesvdaStridedBatched (https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverDN-lt-t-gt-gesvda), which is optimized for tall skinny matrix. You may want to try that in your CUDA program, and see if it gives better performance.

Thank you very much! I have done some experiments, and your answer is exactly the best answer for my problem. The issue #47953 you maintain also helps me a lot.
I did notice the function before, but didn't get the meaning of “strided” at that time... Still a lot to learn in CUDA.

**Comment by ruochiz** [comment]
> 

I encountered similar speed issues when trying to use pytorch to calculate SVD on tall matrices. (In my case it's sth like (bs=300, 900, 100) ) I wonder if gesvdaStridedBatched would be useful in this case, and will this be included in pytorch as a function. I don't have much CUDA programing experiences, so any help would be appreciated. Thanks.

**Comment by Lezcano** [comment]
Benchmark time!

```python
[--------------------- SVD ----------------------]
                             |   cpu    |    cuda 
1 threads: ---------------------------------------
      shape (100, 10, 10)    |    1420  |      406
      shape (300, 900, 100)  |  730200  |  1114000
      shape (1000, 60, 3)    |    3948  |   215200

Times are in microseconds (us)
```

The size in the OP is actually much faster on CUDA than on CPU, so that's one thing less to look at.

When it comes to shape `(300, 900, 100)` in https://github.com/pytorch/pytorch/issues/41306#issuecomment-1001409270, we see that CPU and GPU performance are comparable, so that's not too bad. d`. 

The performance is even worse for the shape `(1000, 60, 3)` from https://github.com/pytorch/pytorch/issues/41306#issuecomment-817419343.

All these things just point to two things
- Performance on GPU vs CPU is not bad in general
- The case for tall / wide matrices is indeed annoying, even more when the matrices are small.

To alleviate the second problem, it would be good to add the backend proposed by @xwang233 to our current heuristics.

**Comment by xwang233** [comment]
Thank you @Lezcano . We're working on to add cusolver gesvdaStridedBatched backend engine to pytorch. We can also provide a SVD heuristic in the best way possible.

However, no heuristic can be perfect. We are also thinking to let user be able to override the backend SVD engine for their use case. Specifically, we want to add a `driver=` kwarg option like `torch.linalg.lstsq` [doc](https://pytorch.org/docs/master/generated/torch.linalg.lstsq.html?highlight=driver). What does the open source community think about this approach?

**Comment by xwang233** [comment]
I've tested the svd performance with the new `gesvda` driver and it looks good on the "skinny" matrices. Tested on V100.

Results:
```
[------------------------------------------------ svd -------------------------------------------------]
                                   |     cpu     |  cuda gesvd  |  cuda gesvdj (default)  |  cuda gesvda
1 threads: ---------------------------------------------------------------------------------------------
      torch.Size([100, 10, 10])    |     1903.5  |    45774.8   |            298.7        |      840.4  
      torch.Size([300, 900, 100])  |  2257043.2  |  2810762.7   |        1608937.1        |    84112.1  
      torch.Size([1000, 60, 3])    |    10318.2  |   225890.6   |         292379.9        |    10484.4  

Times are in microseconds (us).
```

Script:
```python
import torch
from torch.utils.benchmark import Timer, Compare

results = []

def test(_sizes):
    x = torch.randn(*_sizes, dtype=torch.float)
    xcpu = x.cpu()
    xcuda = x.cuda()

    def _subtest(stmt, desc, xcpu=xcpu, xcuda=xcuda):
        t1 = Timer(
            stmt=stmt,
            label='svd',
            sub_label=str(x.size()),
            description=desc,
            globals=dict(globals(), **locals())
            )
        results.append(t1.blocked_autorange())
    
    _subtest('torch.linalg.svd(xcpu)', 'cpu')
    _subtest("torch.linalg.svd(xcuda, driver='gesvd')", 'cuda gesvd')
    _subtest("torch.linalg.svd(xcuda, driver='gesvdj')", 'cuda gesvdj (default)')
    _subtest("torch.linalg.svd(xcuda, driver='gesvda')", 'cuda gesvda')

test((100, 10, 10))
test((300, 900, 100))
test((1000, 60, 3))
Compare(results).print()
```
